include/s3select_json_parser.h

```diff
@@ -0,0 +1,511 @@
+      }
+      return *src_++;
+    }
+
+    void resetBuffer(char* buff, size_t size)
+    {
+      if(!src_)
+      {//first time calling
+	begin_ = buff;
+	src_ = buff;
+	size_ = size;
+	end_= src_ + size_;
+	return;
+      }
+
+      if(!next_src_)
+      {//save the next-chunk that will be used upon parser reaches end of current buffer
+	next_src_ = buff;
+	next_size_ = size;
+      }
+      else
+      {// should not happen
+	std::cout << "can not replace pointers!!!" << std::endl;//TODO exception
+	return;
+      }
+    }
+
+    void saveRemainingBytes()
+    {//this routine called per each new chunk
+      //savine the remaining bytes, before its overriden by the next-chunk.
+      size_t copy_left_sz = getBytesLeft(); //should be very small
+      internal_buffer.assign(src_,copy_left_sz);
+      
+      src_ = internal_buffer.data();
+      begin_ = src_;
+      size_ = copy_left_sz;
+      end_= src_ + copy_left_sz;
+    }
+
+    size_t getBytesLeft() { return end_ - src_; }
+
+};
+
+class JsonParserHandler : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, JsonParserHandler> {
+
+  public:
+
+  typedef enum {OBJECT_STATE,ARRAY_STATE} en_json_elm_state_t;
+
+  public:
+
+    std::vector <std::pair < std::string, Valuesax>> key_value_store;
+    std::vector < std::pair < std::string, Valuesax>> filtermap;
+    std::vector < std::pair < std::string, Valuesax>> exactfilter;
+    std::vector <std::string> filter_pattern{};
+    std::vector <std::string> filter_exact{};
+    std::vector <std::string> filter_prefix_suffix{};
+    std::vector <std::string> prefix{};
+    std::vector <std::string> suffix{};
+    bool flag{};
+    Valuesax value;
+    ChunksStreamer stream_buffer;
+    bool init_buffer_stream;
+    rapidjson::Reader reader;
+    std::vector<en_json_elm_state_t> json_element_state;
+    std::string m_result;//debug purpose
+    std::vector<std::string> key_path;
+    std::string m_filter_result{};
+    std::string m_filter_exact_result{};
+
+    JsonParserHandler() : init_buffer_stream(false)
+    {}
+
+    std::string get_key_path()
+    {//for debug
+	std::string res;
+	for(const auto & i: key_path)
+	{
+	  res.append(i);
+	  res.append(std::string("/"));
+	}
+	return res;
+    }
+
+    void clear_keyValue_store() {
+      key_value_store.clear();
+    }
+
+    void clear_filtermap() {
+      filtermap.clear();
+    }
+
+    void clear_exactfilter() {
+      exactfilter.clear();
+    }
+
+    std::vector < std::pair < std::string, Valuesax>> get_keyValue_store() {
+      return key_value_store;
+    }
+
+    std::vector < std::pair < std::string, Valuesax>> get_filtervalue() {
+      return filtermap;
+    }
+
+    std::vector < std::pair < std::string, Valuesax>> get_exactfiltervalue() {
+      return exactfilter;
+    }
+
+    void dec_key_path()
+    {
+      if(json_element_state.back() != ARRAY_STATE)
+	{
+	  if(key_path.size() != 0)
+	    key_path.pop_back();
+	}
+    }
+
+    void push_new_key_value(Valuesax& v)
```

Comment:
the `push_new_key_value` is being called per each key/value discovered by the parser.
It's the place for filtering the projection-key-path and the where-key-path. (eg:   `select _1.e.f.g   from a.b.c.d`)
it should be an exact match.